### PR TITLE
Improved transaction size calculation.

### LINKED ISSUE #95: Improved fee calculation
Raw transactions contain several compactsize unsigned integers. These are a variable width datatype, however the transaction size function "estimates" the maximum width for these fields. Instead we should calculate the exact size, as we have all the necessary information to do so.

https://github.com/moneybutton/bsv/blob/master/lib/transaction/transaction.js#L909
https://github.com/moneybutton/bsv/blob/master/lib/transaction/transaction.js#L85

### PR DESCRIPTION
**Pull request by Breavyn**
Fixes #95. In my brief tests this yields the exact size of the transaction. The old calculation was actually incorrect, as it didn't account for the 8 bytes that make up the value field of an output.

### CODE REVIEW

**Comment by ryanxcharles** [comment]
SGTM. Please fix the tests.

**Comment by Breavyn** [comment]
Found some more problems with how fees are calculated with the change address function. Need to get that worked out before fixing the tests and merging this.

**Comment by ryanxcharles** [comment]
OK. Please also fix the tests for your new changes. You can run the tests on your local machine to debug. "mocha" or "npm test" or "yarn test".

**Comment by ryanxcharles** [comment]
You are making a lot of changes to the fee calculations. I do recall the estimators in here are broken, which is why we do not use them inside Money Button.

Please note the following properties of estimating size and fees:
* You cannot know precisely the size of a transaction before signing because the signatures can vary in size
* Signatures are somewhere in the range of 65 - 70 bytes. I forget the exact range, but you can figure it out based on the possible sizes of the r and s values.
* In order to estimate the size of an input, you must know how many signatures are going to be in there, and then estimate the size of those signatures.
* Therefore you need to know the "type" of the input (p2sh? multisig?)
* And you typically need to assume the maximum size for a signature, roughly 70 bytes

**Comment by Breavyn** [comment]
That's right, I haven't made any changes to how the size of an input is calculated, and as far as I can tell it still estimates worst case (which seems to only add a couple bytes).
4 of the failing tests appear to be purely from hard coding the old fee calculation result into the tests, which is a simple fix. However the "transaction creation test vector case 6" is concerning me, and I will have to look at it properly. Maybe all it needs is an updated rawtx to compare, but it is weird the other 5 cases work fine as is.
https://github.com/moneybutton/bsv/blob/master/test/transaction/transaction.js#L189
https://github.com/moneybutton/bsv/blob/master/test/data/tx_creation.json#L73-L84

**Comment by Breavyn** [comment]
Don't merge yet, still need to fix multisig (non scrip hash) inputs.

**Comment by Breavyn** [comment]
@ryanxcharles Now ready for review!

File: lib/transaction/transaction.js [hunk 1]
```diff
@@ -828,17 +828,22 @@ Transaction.prototype._updateChangeOutput = function () {
...
+  this._addOutput(new Output({
+    script: this._changeScript,
+    satoshis: 0
+  }))
   var available = this._getUnspentValue()
   var fee = this.getFee()
   var changeAmount = available - fee
+  this._removeOutput(this._changeIndex)
+  this._changeIndex = undefined
   if (changeAmount > 0) {
```
Comment (Breavyn) [inline lib/transaction/transaction.js]: This should probably be `if (changeAmount >= 546) {` to prevent creating a change output below the dust limit.

**Comment by mathiasrw** [comment]
Calculating fees is a bitch. So glad you are taking the quest. Best way I found is to make a twin transaction, sign it and find the size and use it as fee for the "real" transaction.

File: lib/transaction/transaction.js [hunk 2]
```diff
@@ -828,17 +828,22 @@ Transaction.prototype._updateChangeOutput = function () {
   if (!_.isUndefined(this._changeIndex)) {
     this._removeOutput(this._changeIndex)
   }
+  this._changeIndex = this.outputs.length
+  this._addOutput(new Output({
+    script: this._changeScript,
+    satoshis: 0
+  }))
```
Comment (ryanxcharles) [inline lib/transaction/transaction.js]: what are you changing here and why?

File: lib/transaction/transaction.js [hunk 3]
```diff
@@ -828,17 +828,22 @@ Transaction.prototype._updateChangeOutput = function () {
...
   var available = this._getUnspentValue()
   var fee = this.getFee()
   var changeAmount = available - fee
-  if (changeAmount > 0) {
+  this._removeOutput(this._changeIndex)
+  this._changeIndex = undefined
+  if (changeAmount >= Transaction.DUST_AMOUNT) {
```
Comment (ryanxcharles) [inline lib/transaction/transaction.js]: what are you changing here and why?

File: lib/transaction/transaction.js [hunk 4]
```diff
@@ -878,8 +883,7 @@ Transaction.prototype.getFee = function () {
  */
 Transaction.prototype._estimateFee = function () {
   var estimatedSize = this._estimateSize()
-  var available = this._getUnspentValue()
-  return Transaction._estimateFee(estimatedSize, available, this._feePerKb)
+  return Math.ceil(estimatedSize / 1000 * (this._feePerKb || Transaction.FEE_PER_KB))
```
Comment (ryanxcharles) [inline lib/transaction/transaction.js]: why remove the estimator for the change output? is that related to your above changes?

File: lib/transaction/transaction.js (hunk 2 above)
Comment (Breavyn) [reply lib/transaction/transaction.js]: Adding a dummy change output with 0 value. This is used for estimating transaction size accurately, without altering the transactions unspent value.

File: lib/transaction/transaction.js (hunk 3 above)
Comment (Breavyn) [reply lib/transaction/transaction.js]: Remove the dummy output created above. Now we check if there was any change left over. If there is enough change to cover the dust limit, we create the real change output.

File: lib/transaction/transaction.js (hunk 4 above)
Comment (Breavyn) [reply lib/transaction/transaction.js]: We no longer need a change estimator, because we are using a dummy change output from above.

**Comment by Breavyn** [comment]
@ryanxcharles any further thoughts on this?

**Comment by ryanxcharles** [comment]
I'd rather not have to work out whether this is correct or not. You should add more tests that show what the size of different parts should be, with comments explaining why the sizes are what they are, and showing that they come out correct.

File: lib/transaction/transaction.js (hunk 3 above)
Comment (mathiasrw) [reply lib/transaction/transaction.js]: Neat, so doing `.change(adr)` will be ignored if the change value is below dust - right?

**Comment by mathiasrw** [comment]
@Breavyn Thank you so much for this PR.

File: lib/transaction/transaction.js (hunk 3 above)
Comment (Breavyn) [reply lib/transaction/transaction.js]: @mathiasrw essentially yes. Next move is cleaning up datapay, it has some dangerous hacks to work around all this.
